### PR TITLE
New features: override version, detect git URLs, …

### PR DESCRIPTION
**Pull request by iarspider**
specify tags/branches/commits

### CODE REVIEW

**Comment by tldahlgren** [comment]
Can you add unit tests related to the new functionality?  That will bring patch coverage up.  

Not sure if you can see it, but the diffs are at https://app.codecov.io/gh/spack/spack/compare/30921/diff.

**Comment by iarspider** [comment]
@tldahlgren I hope codecov will be happy this time...

**Comment by iarspider** [comment]
@tldahlgren gentle ping

**Comment by tldahlgren** [comment]
> @tldahlgren gentle ping

IIRC I had brought some of these changes up quite a while ago so I'm pinging those who objected at the time to see if they are okay with this approach.

**Comment by tldahlgren** [comment]
> @tldahlgren gentle ping

IIRC I had brought some of these changes up quite a while ago so I'm pinging those who objected at the time to see if they are okay with this approach.

File: lib/spack/spack/cmd/create.py [hunk 1]
```diff
@@ -283,10 +293,8 @@ class RacketPackageTemplate(PackageTemplate):
     """Provides approriate overrides for Racket extensions"""
     base_class_name = 'RacketPackage'
 
-    url_line = """\
-    # FIXME: set the proper location from which to fetch your package
-    git      = "git@github.com:example/example.git"
-    """
+    url_line = "    # FIXME: set the proper location from which to fetch your package"
+    git_line = "    git      = 'git@github.com:example/example.git'"
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: More than 99% of the packages that define the `git` property use the `https` protocol so this statement would be:

```suggestion
    git_line = "    git      = 'https://github.com/example/example.git'"
```

Of those that don't, the protocol is`ssh:` or

```suggestion
    git_line = "    git      = 'ssh://git@github.com/example/example.git'"
```

File: lib/spack/spack/cmd/create.py [hunk 2]
```diff
@@ -125,11 +129,17 @@ def install(self, spec, prefix):
...
 
     url_line = '    url      = "{url}"'
+    git_line = '    git      = "{git_url}"'
 
-    def __init__(self, name, url, versions):
+    def __init__(self, name, url, versions, git_url=None):
         super(PackageTemplate, self).__init__(name, versions)
 
         self.url_def = self.url_line.format(url=url)
+        if git_url is not None:
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: Perhaps better to also not allow a blank string:

```suggestion
        if git_url:
```

File: lib/spack/spack/cmd/create.py [hunk 3]
```diff
@@ -125,11 +129,17 @@ def install(self, spec, prefix):
...
+            self.git_def = self.git_line.format(git_url=git_url)
+        else:
+            self.git_def = '    # FIXME: add git repository if it exists\n'\
+                           '    # git = git@example.com:example/example.git'
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: See suggested options for line 297.

File: lib/spack/spack/cmd/create.py [hunk 4]
```diff
@@ -637,6 +645,31 @@ def setup_parser(subparser):
...
+                                "Not guaranteed to be reproducible, use "
+                                "`--commit` when possible. "
+                                "Only used for git URLs.",
+                           nargs='+')
+    subparser.add_argument('-T', '--tag',
+                           help="specify tag(s) of git repository. "
+                                "Separate multiple tags with space. "
+                                "Not guaranteed to be reproducible, "
+                                "use `--commit` when possible. "
+                                "Only used for git URLs.",
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: This guidance doesn't apply to the `tags` option:

```suggestion
                                "Separate multiple tags with space.",
```

File: lib/spack/spack/cmd/create.py [hunk 5]
```diff
@@ -637,6 +645,31 @@ def setup_parser(subparser):
...
+        '-V', '--version',
+        help='override derived package version')
+    subparser.add_argument('-B', '--branch',
+                           help="specify branch(es) of git repository. "
+                                "Separate multiple branches with space. "
+                                "Not guaranteed to be reproducible, use "
+                                "`--commit` when possible. "
+                                "Only used for git URLs.",
+                           nargs='+')
+    subparser.add_argument('-T', '--tag',
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: Spack actually has a tag argument "standard" from `cmd/common/arguments.py`.  It is used in the following commands:

- `spack external`
- `spack find`

and added with

```
    spack.cmd.common.arguments.add_common_arguments(sub parser, ["tags"])
```

File: lib/spack/spack/cmd/create.py [hunk 6]
```diff
@@ -637,6 +645,31 @@ def setup_parser(subparser):
...
         help="don't ask which versions to checksum")
+    subparser.add_argument(
+        '-g', '--git', action='store', nargs='?', default='#AUTO-GIT-URL#',
+        help="use git to download source from repository passed in url argument")
+    subparser.add_argument(
+        '-V', '--version',
+        help='override derived package version')
+    subparser.add_argument('-B', '--branch',
+                           help="specify branch(es) of git repository. "
+                                "Separate multiple branches with space. "
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: Multiple `-B` (or `--branch`) options should be allowed in the same way we support `tags`.

An example where the space-separated approach doesn't work as expected:

```
$ spack create -n mpileaks-git -g https://github.com/LLNL/mpileaks.git -B master https://github.com/LLNL/mpileaks/releases/download/v1.0/mpileaks-1.0.tar.gz
```

File: lib/spack/spack/cmd/create.py [hunk 7]
```diff
@@ -637,6 +645,31 @@ def setup_parser(subparser):
...
+                                "Only used for git URLs.",
+                           nargs='+')
+    subparser.add_argument('-T', '--tag',
+                           help="specify tag(s) of git repository. "
+                                "Separate multiple tags with space. "
+                                "Not guaranteed to be reproducible, "
+                                "use `--commit` when possible. "
+                                "Only used for git URLs.",
+                           nargs='+')
+    subparser.add_argument('-C', '--commit',
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: This has the same problem as the other space-separated options.

File: lib/spack/spack/cmd/create.py [hunk 8]
```diff
@@ -771,26 +804,84 @@ def get_name(args):
...
+    git = None
+
+    # No source and no git urls were provided
+    if args.url is None and args.git is None:
+        return url, git
+
+    # Git url not set explicitly
+    if (args.git == '#AUTO-GIT-URL#' and is_git_url(args.url)) or args.git is None:
+        git = args.url
+        return url, git
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: Would it be helpful to return the following (so the `url` is not the example?

```suggestion
        return args.url, git
```

File: lib/spack/spack/cmd/create.py [hunk 9]
```diff
@@ -818,10 +909,56 @@ def get_versions(args, name):
...
     # Default guesser
     guesser = BuildSystemGuesser()
 
     valid_url = True
+
+    url, git = get_url_and_git(args)
+
+    has_git_option = args.commit is not None or \
+        args.tag is not None or \
+        args.branch is not None
```
Comment (tldahlgren) [inline lib/spack/spack/cmd/create.py]: This could be tightened:

```suggestion
    has_git_option = args.commit or args.tag or args.branch
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
The main issue is to not use space separation for multi-valued options.

**Review by tldahlgren — APPROVED** [review]

**Pull request closed without merging**